### PR TITLE
Unit test coverage for WaveGlow -> ONNX conversion

### PR DESCRIPTION
**Pull request by drnikolaev**
This is #581 v2.
No structural changes except calling `self.waveglow.forward` in `WaveGlowInferNM.forward` (this is a must). All other changes are to make the code tracer friendly.
As of `__call__` hacking - I don't see any difference. I just moved it to `try` block since we rely on complimentary `finally`. Also, the `__old_call__` body was incorrect, now we just copy it as is.

### CODE REVIEW

File: nemo/collections/tts/waveglow_modules.py [hunk 1]
```diff
@@ -207,16 +207,15 @@ def denoise(self, audio, strength=0.1):
         audio_denoised = librosa.core.istft(audio_spec_denoised * audio_angles)
         return audio_denoised, audio_spec_denoised
 
-    def forward(self, mel_spectrogram):
+    def forward(self, *args):
```
Comment (okuchaiev) [inline nemo/collections/tts/waveglow_modules.py:210]: is it possible to put named arguments (if necessary with default values back instead of "*argc")?

File: tests/unit/core/test_deploy_export.py [hunk 2]
```diff
@@ -350,3 +341,29 @@ def test_hf_bert(self, tmpdir, df_type):
...
+    @pytest.mark.run_only_on('GPU')
+    @pytest.mark.parametrize("df_type", [DF.ONNX, DF.TORCHSCRIPT, DF.PYTORCH])
+    def test_waveglow(self, tmpdir, df_type):
+        url = "https://api.ngc.nvidia.com/v2/models/nvidia/waveglow_ljspeech/versions/2/files/WaveGlowNM.pt"
+        ptfile = "./WaveGlowNM.pt"
+        if not Path(ptfile).is_file():
+            urllib.request.urlretrieve(url, ptfile)
+
+        module = nemo_tts.WaveGlowInferNM(sample_rate=22050)
+        module.load_state_dict(torch.load(ptfile))
```
[the remark this answers is not in the document]
Comment (drnikolaev) [reply tests/unit/core/test_deploy_export.py:355]: done

File: tests/unit/core/test_deploy_export.py [hunk 3]
```diff
@@ -350,3 +341,29 @@ def test_hf_bert(self, tmpdir, df_type):
...
+        module.eval()
```
[the remark this answers is not in the document]
Comment (drnikolaev) [reply tests/unit/core/test_deploy_export.py:356]: No, `eval()` turns off training mode and we need that immediately

File: tests/unit/core/test_deploy_export.py [hunk 4]
```diff
@@ -350,3 +341,29 @@ def test_hf_bert(self, tmpdir, df_type):
...
+        module.cuda()
```
[the remark this answers is not in the document]
Comment (drnikolaev) [reply tests/unit/core/test_deploy_export.py:357]: Yes. Removed

File: nemo/collections/tts/waveglow_modules.py (hunk 1 above)
Comment (drnikolaev) [reply nemo/collections/tts/waveglow_modules.py:210]: Done

File: nemo/collections/tts/waveglow_modules.py [hunk 5]
```diff
@@ -161,7 +161,7 @@ def output_ports(self):
         return {"audio": NeuralType(('B', 'T'), AudioSignal(freq=self.sample_rate))}
 
     def __str__(self):
-        return "WaveGlowNM"
+        return "WaveGlowInferNM"
```
Comment (blisc) [inline nemo/collections/tts/waveglow_modules.py:164]: Is the name change necessary?

Comment (drnikolaev) [reply nemo/collections/tts/waveglow_modules.py:164]: It helps Tracer to say the truth about what class it fails in.

**Comment by blisc** [comment]
By changing the name, it breaks our current tts_infer script inside examples.
So now our examples don't work.

File: nemo/collections/tts/waveglow_modules.py (hunk 5 above)
Comment (blisc) [reply nemo/collections/tts/waveglow_modules.py:164]: Is it necessary for converting waveglow into an onnx format?
If not, I am changing it back.

**Comment by drnikolaev** [comment]
@blisc yes, please change it back.

File: nemo/collections/tts/waveglow_modules.py (hunk 5 above)
Comment (drnikolaev) [reply nemo/collections/tts/waveglow_modules.py:164]: @blisc yes, please change it back.

File: nemo/collections/tts/waveglow_modules.py [hunk 6]
```diff
@@ -207,16 +207,15 @@ def denoise(self, audio, strength=0.1):
...
             logging.info("remove WN")
-            self.waveglow = self.waveglow.remove_weightnorm(self.waveglow)
+            self.waveglow = remove_weightnorm(self.waveglow)
             self._removed_weight_norm = True
         if self.training:
             raise ValueError("You are using the WaveGlow Infer Neural Module in training mode.")
         with torch.no_grad():
-            audio = self.waveglow.infer(mel_spectrogram, sigma=self._sigma)
-        return audio
+            return self.waveglow.forward((mel_spectrogram, z,))[0]
```
Comment (blisc) [inline nemo/collections/tts/waveglow_modules.py:218]: @drnikolaev, why did you add z to the forward method of `WaveglowInferNM` and change it from waveglow.infer to waveglow.forward?

Comment (blisc) [reply nemo/collections/tts/waveglow_modules.py:218]: @titu1994

Comment (titu1994) [reply nemo/collections/tts/waveglow_modules.py:218]: @drnikolaev Is this required for ONNX or TRT? If its for conversion compat, can I add a flag that uses infer by default, but switches to this forward? Also, the input_ports needs to be modified to support the `z` variable in such a case.

Comment (drnikolaev) [reply nemo/collections/tts/waveglow_modules.py:218]: It was made to avoid tracer error which was unrecoverable. You can try to undo it and see yourself.

Comment (titu1994) [reply nemo/collections/tts/waveglow_modules.py:218]: So will the flag technique support both use cases ? Or will tracer (is this the JIT trace of pytorch?) complain that execution paths have different inputs and outputs ? Also, how do we supply `z` to the neural module as it is currently ?

Comment (drnikolaev) [reply nemo/collections/tts/waveglow_modules.py:218]: I'll need some time to reproduce the error, please hold on

File: tests/unit/core/test_deploy_export.py [hunk 7]
```diff
@@ -350,3 +341,28 @@ def test_hf_bert(self, tmpdir, df_type):
...
+        mel = torch.randn(1, 80, 96).cuda()
+        stride = 256  # value from waveglow upsample
+        n_group = 8
+        z_size2 = (mel.size(2) * stride) // n_group
+        z = torch.randn(1, z_size2).cuda()
+
+        input_example = OrderedDict([("mel_spectrogram", mel), ("z", z)])
+        tmp_file_name = str(tmpdir.mkdir("export").join("waveglow"))
+
+        self.__test_export_route(module=module, out_name=tmp_file_name, mode=df_type, input_example=input_example)
```
Comment (blisc) [inline tests/unit/core/test_deploy_export.py:368]: I am now very confused by this PR. Are you sure that you are exporting the inference route when you do this?
